### ISSUE TITLE
Show badge with unread counter (opt-in)

### ISSUE THREAD

**Issue by brunolemos** [devhubapp/devhub#90]
Show badge on browser icon, desktop icons and mobile icons.
Only for the columns that the user wants (add the option).

> Note: Code for this is partially ready so no need to contribute on this one

**Comment by panva** [comment]
@brunolemos how about Menubar mode badge in the status bar?

**Comment by dkniffin** [comment]
Menubar badge would be 💯. I've been using [Gitify](https://github.com/manosim/gitify) for a long time, and it changes the color when there's unread notifications, which works great. But knowing the number too somehow would be icing on that cake.

**Comment by dkniffin** [comment]
I might actually take a stab at implementing that. I'm familiar with the react-native-web architecture, but not the specifics of this project. Is [this](https://github.com/devhubapp/devhub/blob/master/packages/desktop/src/tray.ts#L17-L22) the right place for that change?

Also, for the number, the only way I can think to do it is to have different images for each number. Maybe created dynamically as discussed here: https://github.com/electron/electron/issues/7322

**Comment by brunolemos** [comment]
@dkniffin yes that would be the file!

I think the hardest part will be that each column needs to have a config to enable or disable the badge; but we can think of a simpler solution to start.

I recently created a method called `getItemsFilterMetadata` that returns the number of unread items for a specific column and that was a requirement for the badge feature, so it will come handy.

https://github.com/devhubapp/devhub/blob/e096f471886c9f14bf73382820f53e7c0161ef5d/packages/components/src/components/columns/ColumnOptions.tsx#L525-L533

https://github.com/devhubapp/devhub/blob/e096f471886c9f14bf73382820f53e7c0161ef5d/packages/components/src/components/columns/ColumnOptions.tsx#L625

The code above shows the number `49` at the screenshot below: 

![image](https://user-images.githubusercontent.com/619186/60130971-b0357180-976e-11e9-86bd-7c43e88c60f6.png)

**Comment by johnletey** [comment]
@dkniffin are you going to just work on the icon badge counter? if so, I can have a go at the "unread dot" on sidebar items (kind of like Slack and Discord)

**Comment by dkniffin** [comment]
@johnletey Yeah, if I do anything on here, it would be the menubar indicator

**Comment by johnletey** [comment]
@dkniffin Wonderful ... I will get started on the "unread dot" feature now!

**Comment by dkniffin** [comment]
@brunolemos I'm trying to get the menubar/tray indicator working. I know how to render it, but I'm having a hard time figuring out how to get the data needed to determine which state to render it in. You mentioned `getItemsFilterMetadata`, but I'm not sure how to use that in the `Tray.ts` file, since I need access to a column, but I don't know which column is being displayed. Thoughts?

**Comment by brunolemos** [comment]
@dkniffin the `tray.ts` file is the last step, everything else will happen inside the `components` package.

Check the file `Sidebar.tsx`, it is a mess but it has everything you need there. It lists all the columns and show an unread indicator.

I think we need a React Context, e.g. `UnreadWatcherContext.tsx`, that will be responsible to check the unread indicator for all columns. If there are columns with unread items, let’s sum them an communicate Electron about it. 

To send messages to Electron, first check that it is electron an electron app (`Platform.isElectron && window.ipc`) and send the event (e.g. `window.ipc.send(‘unread-counter’, numberOfUnreadItems)`).

You need to listen to this `unread-counter` event on Electron (file `desktop/src/index.ts`), with a `app.on(‘unread-counter’, …` event listener). And then change the menubar icon badge (`tray.yourMethodHere(counter)`).

**Comment by dkniffin** [comment]
Ok, thanks. That makes sense. I'll give that a shot

**Comment by brunolemos** [comment]
You can use this to get all the columns:

`const columns = useReduxState(selectors.columnsArrSelector)`

And then for each column you do something like this to get the unread count for each one:

https://github.com/devhubapp/devhub/blob/838840a272cb33b1a3229397740369e8fbb9537a/packages/components/src/components/layout/Sidebar.tsx#L614-L635

**Comment by dkniffin** [comment]
@brunolemos These context objects are unfamiliar to me. Are these the same as [React Contexts](https://reactjs.org/docs/context.html) or is this a different thing?

**Comment by brunolemos** [comment]
@dkniffin yes, normal React Context, like this: https://github.com/devhubapp/devhub/blob/master/packages/components/src/components/context/ColumnWidthContext.tsx

**Comment by dkniffin** [comment]
@brunolemos Sorry I keep pestering you with questions about this. I think I've got the context set up correctly, and I understand how to pass the event over to electron, but the piece I'm missing is where to actually do that passing. Does that belong in `MainScreen.tsx` or somewhere else?

**Comment by brunolemos** [comment]
Hm you can do it directly inside the context provider for now, or create a new file (e.g. `AppIconBadge.tsx`) and include it at the MainScreen render (`<AppIconBadge />`). And then inside this file read the context value (that should return the number of unread items) and update the app icon accordingly.

**Comment by brunolemos** [comment]
This is now available on [v0.93.0](https://github.com/devhubapp/devhub/releases/tag/v0.93.0)! For web and desktop.

To show an unread badge on mobile there are two possible ways:
1. Run a script on app background every X minutes (has limitations: app needs to stay minimized, minimum 15 minute interval to update, etc)
2. Move this responsibility to the server (probably would not scale well, imagine the server having to do thousands of github API requests per minute)

#2 would provide a better UX but not sure if it's feasible, github would block the server probably. Help wanted on #1 for now!

**Comment by brunolemos** [comment]
To show an unread badge on mobile there are two possible ways:
1. Run a script on app background every X minutes (has limitations: app needs to stay minimized, minimum 15 minute interval to update, etc)
2. Move this responsibility to the server (probably would not scale well, imagine the server having to do thousands of github API requests per minute)

N2 would provide a better UX but not sure if it's feasible, github would block the server probably. Help wanted on N1 for now!

**Comment by satyaarthchhabra** [comment]
Hi @brunolemos, I think the issue is solved now hence we can close this issue as well, if there is something left , I would love to contribute please tell me
